controllers/clustertopology_controller_compute.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// getNestedRef returns the ref value of a nested field.
+// NOTE: This function assumes the reference either exists in the CRD, and in that case the value is required,
+// or it does not exists in the CRD.
+func getNestedRef(obj *unstructured.Unstructured, fields ...string) *corev1.ObjectReference {
```

Comment:
In case one of the fields is not set, we could either return:
* a partial ObjectReference (current impl)
* nil

I wonder what's easier to handle for the consumers. A concrete example:

I try to retrieve the optional InfrastructureMachineTemplate from KubeadmControlPlane via getNestedRef

The following could happen:
* ObjectRef not set => `getNestedRef` retuns nil => I assume the ref doesn't exist on the ControlPlane
* ObjectRef is set => `getNestedRef` returns an ObjectReference => I handle the MachineTemplate
* ObjectRef is partially set => Now we have two options for the implementation:
    * => `getNestedRef` returns nil => I assume the ref doesn't exist on the ControlPlane
    * => `getNestedRef` returns a partial ObjectReference => I have to either just use the ObjectReference and the Get will fail or I have to validate if I got a "full" ObjectReference and handle it accordingly

So overall: I just wanted to ask if we just want to interpret a partial ObjectReference the same way as a non-existing one (consistently), then we wouldn't have to do anything on the consumer side when we return nil instead of a partial ObjectReference.

Reply:
FYI we are enforcing reference should be valid/complete in https://github.com/kubernetes-sigs/cluster-api/issues/4940

Assuming the above context, this func should get what it founds and we can reasonably assume the value is correct and complete; If we are going to find out this assumption too week while implementing consumers in the upcoming PRs, we can iterate and improve accordingly

Reply:
I think #4940 does not include the ref in ControlPlane. But I think that's the responsibility of the control plane provider and otherwise they will just get an error here which is the best we can do in the controller code anyway.

So agree, let's keep it as is.

Related follow up question: Maybe I'm missing something but could it be that KCP only validates 3/4 fields of the ObjectReference?
https://github.com/kubernetes-sigs/cluster-api/blob/master/controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_webhook.go#L280-L309
(our 4 relevant fields are all optional in the OpenAPI schema)

In case I'm right, I can open a PR to add the validation for `Name`

Reply:
Validating name was should go in #4940; please comment there